### PR TITLE
Handle sequential inserts with same position

### LINKED ISSUE #12526: LSP: tsserver "add missing import" code action results in syntax error
- `nvim --version`: NVIM v0.5.0-721f69c4a
- language server name/version: tsserver
- Operating system/version: macOS Catalina

<details>
<summary>nvim -c ":checkhealth nvim nvim_lsp"</summary>

health#nvim#check
========================================================================
## Configuration
  - OK: no issues found

## Performance
  - OK: Build type: Release

## Remote Plugins
  - OK: Up to date

## terminal
  - INFO: key_backspace (kbs) terminfo entry: key_backspace=^H
  - INFO: key_dc (kdch1) terminfo entry: key_dc=\E[3~
  - INFO: $COLORTERM='truecolor'

## tmux
  - OK: escape-time: 10ms
  - INFO: $TERM: tmux-256color

health#nvim_lsp#check
========================================================================
## Checking language server protocol configuration
  - INFO: pyls: configuration checked.
  - INFO: jdtls: configuration checked.
  - INFO: tsserver: configuration checked.
  - INFO: rust_analyzer: configuration checked.
  - INFO: metals: configuration checked.

</details>

<details>
<summary>lsp.log</summary>

[ DEBUG ] 2020-06-24T07:09:08+0900 ] ...r/neovim/HEAD-721f69c/share/nvim/runtime/lua/vim/lsp.lua:504 ]  "LSP[tsserver]" "client.request"        1       "textDocument/codeAction"      {  context = {    diagnostics = { {        code = 2304,        message = "Cannot find name 'C'.",        range = {          end = {            character = 10,            line = 5          },          start = {            character = 9,            line = 5          }        },        severity = 1,        source = "typescript",        uri = "file:///Users/kevin/dev/nvim_bug/main.ts"      }, {        code = 6133,        message = "'c' is declared but its value is never read.",        range = {          end = {            character = 7,            line = 5          },          start = {            character = 6,            line = 5          }        },        severity = 4,        source = "typescript",        uri = "file:///Users/kevin/dev/nvim_bug/main.ts"      } }  },  range = {    end = <1>{      character = 9,      line = 5    },    start = <table 1>  },  textDocument = {    uri = "file:///Users/kevin/dev/nvim_bug/main.ts"  }}      <function 1>    1
[ DEBUG ] 2020-06-24T07:09:08+0900 ] ...ovim/HEAD-721f69c/share/nvim/runtime/lua/vim/lsp/rpc.lua:267 ]  "rpc.send.payload"      {  id = 2,  jsonrpc = "2.0",  method = "textDocument/codeAction",  params = {    context = {      diagnostics = { {          code = 2304,          message = "Cannot find name 'C'.",          range = {            end = {              character = 10,              line = 5            },            start = {              character = 9,              line = 5            }          },          severity = 1,          source = "typescript",          uri = "file:///Users/kevin/dev/nvim_bug/main.ts"        }, {          code = 6133,          message = "'c' is declared but its value is never read.",          range = {            end = {              character = 7,              line = 5            },            start = {              character = 6,              line = 5            }          },          severity = 4,          source = "typescript",          uri = "file:///Users/kevin/dev/nvim_bug/main.ts"        } }    },    range = {      end = <1>{        character = 9,        line = 5      },      start = <table 1>    },    textDocument = {      uri = "file:///Users/kevin/dev/nvim_bug/main.ts"    }  }}
[ DEBUG ] 2020-06-24T07:09:08+0900 ] ...ovim/HEAD-721f69c/share/nvim/runtime/lua/vim/lsp/rpc.lua:347 ]  "decoded"       {  id = 2,  jsonrpc = "2.0",  result = { {      arguments = { {          documentChanges = { {              edits = { {                  newText = ",",                  range = {                    end = {                      character = 3,                      line = 2                    },                    start = {                      character = 3,                      line = 2                    }                  }                }, {                  newText = "\n  C",                  range = {                    end = {                      character = 3,                      line = 2                    },                    start = {                      character = 3,                      line = 2                    }                  }                } },              textDocument = {                uri = "file:///Users/kevin/dev/nvim_bug/main.ts",                version = 0              }            } }        } },      command = "_typescript.applyWorkspaceEdit",      title = "Add 'C' to existing import declaration from \"./lib\""    }, {      arguments = { {          documentChanges = { {              edits = { {                  newText = "",                  range = {                    end = {                      character = 10,                      line = 5                    },                    start = {                      character = 8,                      line = 5                    }                  }                } },              textDocument = {                uri = "file:///Users/kevin/dev/nvim_bug/main.ts",                version = 0              }            } }        } },      command = "_typescript.applyWorkspaceEdit",      title = "Remove declaration for: 'C'"    } }}
[ DEBUG ] 2020-06-24T07:09:08+0900 ] ...EAD-721f69c/share/nvim/runtime/lua/vim/lsp/callbacks.lua:257 ]  "default_callback"      "textDocument/codeAction"     {  bufnr = 1,  client_id = 1,  params = { {      arguments = { {          documentChanges = { {              edits = { {                  newText = ",",                  range = {                    end = {                      character = 3,                      line = 2                    },                    start = {                      character = 3,                      line = 2                    }                  }                }, {                  newText = "\n  C",                  range = {                    end = {                      character = 3,                      line = 2                    },                    start = {                      character = 3,                      line = 2                    }                  }                } },              textDocument = {                uri = "file:///Users/kevin/dev/nvim_bug/main.ts",                version = 0              }            } }        } },      command = "_typescript.applyWorkspaceEdit",      title = "Add 'C' to existing import declaration from \"./lib\""    }, {      arguments = { {          documentChanges = { {              edits = { {                  newText = "",                  range = {                    end = {                      character = 10,                      line = 5                    },                    start = {                      character = 8,                      line = 5                    }                  }                } },              textDocument = {                uri = "file:///Users/kevin/dev/nvim_bug/main.ts",                version = 0              }            } }        } },      command = "_typescript.applyWorkspaceEdit",      title = "Remove declaration for: 'C'"    } }}[ DEBUG ] 2020-06-24T07:09:16+0900 ] ...r/neovim/HEAD-721f69c/share/nvim/runtime/lua/vim/lsp.lua:504 ]  "LSP[tsserver]" "client.request"        1       "workspace/executeCommand"     {  arguments = { {      documentChanges = { {          edits = { {              newText = ",",              range = {                end = {                  character = 3,                  line = 2                },                start = {                  character = 3,                  line = 2                }              }            }, {              newText = "\n  C",              range = {                end = {                  character = 3,                  line = 2                },                start = {                  character = 3,                  line = 2                }              }            } },          textDocument = {            uri = "file:///Users/kevin/dev/nvim_bug/main.ts",            version = 0          }        } }    } },  command = "_typescript.applyWorkspaceEdit",  title = "Add 'C' to existing import declaration from \"./lib\""}        <function 1>    1[ DEBUG ] 2020-06-24T07:09:16+0900 ] ...ovim/HEAD-721f69c/share/nvim/runtime/lua/vim/lsp/rpc.lua:267 ]  "rpc.send.payload"      {  id = 3,  jsonrpc = "2.0",  method = "workspace/executeCommand",  params = {    arguments = { {        documentChanges = { {            edits = { {                newText = ",",                range = {                  end = {                    character = 3,                    line = 2                  },                  start = {                    character = 3,                    line = 2                  }                }              }, {                newText = "\n  C",                range = {                  end = {                    character = 3,                    line = 2                  },                  start = {                    character = 3,                    line = 2                  }                }              } },            textDocument = {              uri = "file:///Users/kevin/dev/nvim_bug/main.ts",              version = 0            }          } }      } },    command = "_typescript.applyWorkspaceEdit",    title = "Add 'C' to existing import declaration from \"./lib\""  }}[ DEBUG ] 2020-06-24T07:09:16+0900 ] ...ovim/HEAD-721f69c/share/nvim/runtime/lua/vim/lsp/rpc.lua:347 ]  "decoded"       {  id = 0,  jsonrpc = "2.0",  method = "workspace/applyEdit",  params = {    edit = {      documentChanges = { {          edits = { {              newText = ",",              range = {                end = {                  character = 3,                  line = 2                },                start = {                  character = 3,                  line = 2                }              }            }, {              newText = "\n  C",              range = {                end = {                  character = 3,                  line = 2                },                start = {                  character = 3,                  line = 2                }              }            } },          textDocument = {            uri = "file:///Users/kevin/dev/nvim_bug/main.ts",            version = 0          }        } }    }  }}[ DEBUG ] 2020-06-24T07:09:16+0900 ] ...r/neovim/HEAD-721f69c/share/nvim/runtime/lua/vim/lsp.lua:352 ]  "server_request"        "workspace/applyEdit"   {  edit = {    documentChanges = { {        edits = { {            newText = ",",            range = {              end = {                character = 3,                line = 2              },              start = {                character = 3,                line = 2              }            }          }, {            newText = "\n  C",            range = {              end = {                character = 3,                line = 2              },              start = {                character = 3,                line = 2              }            }          } },        textDocument = {          uri = "file:///Users/kevin/dev/nvim_bug/main.ts",          version = 0        }      } }  }}
[ DEBUG ] 2020-06-24T07:09:16+0900 ] ...r/neovim/HEAD-721f69c/share/nvim/runtime/lua/vim/lsp.lua:355 ]  "server_request: found callback for"    "workspace/applyEdit"[ DEBUG ] 2020-06-24T07:09:16+0900 ] ...EAD-721f69c/share/nvim/runtime/lua/vim/lsp/callbacks.lua:257 ]  "default_callback"      "workspace/applyEdit"   {  client_id = 1,  params = {    edit = {      documentChanges = { {          edits = { {              newText = ",",              range = {                end = {                  character = 3,                  line = 2                },                start = {                  character = 3,                  line = 2                }              }            }, {              newText = "\n  C",              range = {                end = {                  character = 3,                  line = 2                },                start = {                  character = 3,                  line = 2                }              }            } },          textDocument = {            uri = "file:///Users/kevin/dev/nvim_bug/main.ts",            version = 0          }        } }    }  }}[ DEBUG ] 2020-06-24T07:09:16+0900 ] ...r/neovim/HEAD-721f69c/share/nvim/runtime/lua/vim/lsp.lua:587 ]  "on_lines"      1       4       2       3       4     44       4       { "  B", "  C," }[ DEBUG ] 2020-06-24T07:09:16+0900 ] ...ovim/HEAD-721f69c/share/nvim/runtime/lua/vim/lsp/rpc.lua:278 ]  "rpc.notify"    "textDocument/didChange"        {  contentChanges = { {      text = 'import {\n  A,\n  B\n  C,\n} from "./lib";\n\nconst c: C = {};\n'    } },  textDocument = {    uri = "file:///Users/kevin/dev/nvim_bug/main.ts",    version = 4  }}[ DEBUG ] 2020-06-24T07:09:16+0900 ] ...ovim/HEAD-721f69c/share/nvim/runtime/lua/vim/lsp/rpc.lua:267 ]  "rpc.send.payload"      {  jsonrpc = "2.0",  method = "textDocument/didChange",  params = {    contentChanges = { {        text = 'import {\n  A,\n  B\n  C,\n} from "./lib";\n\nconst c: C = {};\n'      } },    textDocument = {      uri = "file:///Users/kevin/dev/nvim_bug/main.ts",      version = 4    }  }}[ DEBUG ] 2020-06-24T07:09:16+0900 ] ...ovim/HEAD-721f69c/share/nvim/runtime/lua/vim/lsp/rpc.lua:358 ]  "server_request: callback result"       {  result = {    applied = true  },  status = true}[ DEBUG ] 2020-06-24T07:09:16+0900 ] ...ovim/HEAD-721f69c/share/nvim/runtime/lua/vim/lsp/rpc.lua:267 ]  "rpc.send.payload"      {  id = 0,  jsonrpc = "2.0",  result = {    applied = true  }}[ DEBUG ] 2020-06-24T07:09:16+0900 ] ...ovim/HEAD-721f69c/share/nvim/runtime/lua/vim/lsp/rpc.lua:347 ]  "decoded"       {  id = 3,  jsonrpc = "2.0",  result = vim.NIL}[ DEBUG ] 2020-06-24T07:09:16+0900 ] ...EAD-721f69c/share/nvim/runtime/lua/vim/lsp/callbacks.lua:257 ]  "default_callback"      "workspace/executeCommand"    {  bufnr = 1,  client_id = 1}[ DEBUG ] 2020-06-24T07:09:16+0900 ] ...ovim/HEAD-721f69c/share/nvim/runtime/lua/vim/lsp/rpc.lua:347 ]  "decoded"       {  jsonrpc = "2.0",  method = "textDocument/publishDiagnostics",  params = {    diagnostics = { {        code = 1005,        message = "',' expected.",        range = {          end = {            character = 3,            line = 3          },          start = {            character = 2,            line = 3          }        },        severity = 1,        source = "typescript"      }, {        code = 6133,        message = "'A' is declared but its value is never read.",        range = {          end = {            character = 3,            line = 1          },          start = {            character = 2,            line = 1          }        },        severity = 4,        source = "typescript"      }, {        code = 6133,        message = "'B' is declared but its value is never read.",        range = {          end = {            character = 3,            line = 2          },          start = {            character = 2,            line = 2          }        },        severity = 4,        source = "typescript"      }, {        code = 6133,        message = "'c' is declared but its value is never read.",        range = {          end = {            character = 7,            line = 6          },          start = {            character = 6,            line = 6          }        },        severity = 4,        source = "typescript"      } },    uri = "file:///Users/kevin/dev/nvim_bug/main.ts"  }}
[ DEBUG ] 2020-06-24T07:09:16+0900 ] ...r/neovim/HEAD-721f69c/share/nvim/runtime/lua/vim/lsp.lua:343 ]  "notification"  "textDocument/publishDiagnostics"     {  diagnostics = { {      code = 1005,      message = "',' expected.",      range = {        end = {          character = 3,          line = 3        },        start = {          character = 2,          line = 3        }      },      severity = 1,      source = "typescript"    }, {      code = 6133,      message = "'A' is declared but its value is never read.",      range = {        end = {          character = 3,          line = 1        },        start = {          character = 2,          line = 1        }      },      severity = 4,      source = "typescript"    }, {      code = 6133,      message = "'B' is declared but its value is never read.",      range = {        end = {          character = 3,          line = 2        },        start = {          character = 2,          line = 2        }      },      severity = 4,      source = "typescript"    }, {      code = 6133,      message = "'c' is declared but its value is never read.",      range = {        end = {          character = 7,          line = 6        },        start = {          character = 6,          line = 6        }      },      severity = 4,      source = "typescript"    } },  uri = "file:///Users/kevin/dev/nvim_bug/main.ts"}

</details>

#### Setup
Not NORC, but can be reproduced with this minimal `init.vim` (using vim-plug to install the nvim-lsp plugin):
```
call plug#begin(stdpath('data') . '/plugged')
Plug 'neovim/nvim-lsp'
call plug#end()

lua << EOF
local nvim_lsp = require'nvim_lsp'
nvim_lsp.tsserver.setup{}
EOF
```

Also requires the tsserver executable:
```
npm install -g typescript-language-server
```

Then create these two files in a new directory:

`lib.ts`, with contents
```typescript
export interface A {}
export interface B {}
export interface C {}
```

`main.ts`, with contents
```typescript
import {
  A,
  B
} from "./lib";

const c: C = {};
```

Finally, create a `package.json` file so nvim-lsp will recognize the project and use tsserver (can do `npm init` and just use the defaults).

#### Reproducing

1. Open `main.ts` using the minimal `init.vim` written above
1. Put your cursor on the capital `C` in the last line. Note that `C` isn't yet imported from `./lib`, so we want to use a code action to import it.
1. Do `:lua vim.lsp.buf.code_action()<CR>`
1. Choose the code action `Add 'C' to existing import declaration from "./lib"`

### Actual behaviour

The imports are changed to

```typescript
import {
  A,
  B
  C,
} from "./lib";
```

(which is a syntax error; there should be a comma after `B`)

### Expected behaviour

The imports are changed to

```typescript
import {
  A,
  B,
  C
} from "./lib";
```

### More info

`git bisect` suggests that this started happening with https://github.com/neovim/neovim/commit/44fe8828f06a22bc9aa3617a6fd8aae447a838de; it was working as expected before that. The comments on that PR (https://github.com/neovim/neovim/pull/12434) suggest that before that, nvim's LSP was applying text edits in an incorrect order...so if nvim's behavior is in fact correct according to the spec now, then it's possible that this is a bug in tsserver. Looking at the LSP logs, though, the relevant text edits seem to be of this form:

```
{
  id = 2,
  jsonrpc = "2.0",
  result = {
    {
      arguments = {
        {
          documentChanges = {
            {
              edits = {
                {
                  newText = ",",
                  range = {
                    end = {
                      character = 3,
                      line = 2
                    },
                    start = {
                      character = 3,
                      line = 2
                    }
                  }
                },
                {
                  newText = "\n  C",
                  range = {
                    end = {
                      character = 3,
                      line = 2
                    },
                    start = {
                      character = 3,
                      line = 2
                    }
                  }
                }
              },
  ...
```

I don't know too much about the LSP spec, but it looks like two text edits with the same start position: first add a comma, and then add a newline and `C`. So it seems like applying them in that order should result in the expected code...I'm not sure why it seems to be happening backwards. The particular conditions (two textedits with the same start position, and something possibly happening with the ordering) do seem to have something to do with https://github.com/neovim/neovim/commit/44fe8828f06a22bc9aa3617a6fd8aae447a838de, though.

### PR DESCRIPTION
**Pull request by ndreas**
The change introduced in #12434 was a fix for the case when a remove was followed by an insert in the same position, which came in the wrong order before #12434. This, however had the effect that sequential inserts with the same position broke, see #12526.

According to the spec:

> If multiple inserts have the same position, the order in the array defines the order in which the inserted strings
appear in the resulting text.

To fix both the case in #12434 and the spec case, this change keeps track of any previous inserts and collapses them into a single change, thus making sure that the inserted strings appear in the correct order.

I am unsure if this is a functionality that is desired. I interpret the spec to actually make remove followed by insert an invalid operation, but it seems like a reasonable thing to do. `gopls` does this when organzing imports.

This updates and adds test cases for both scenarios, and should fix #12526

### CODE REVIEW

File: test/functional/plugin/lsp_spec.lua [hunk 1]
```diff
@@ -876,6 +853,56 @@ describe('LSP', function()
...
+          'newline';
+          'Second line of text';
+          'Third line of text';
+          'Fourth line of text';
+          'å å ɧ 汉语 ↥ 🤦 🦄';
+        }, buf_lines(1))
+      end)
+
+      -- Not really valid according to the spec, but some language servers send
+      -- edits like this
```
Comment (teto) [inline test/functional/plugin/lsp_spec.lua:880]: which ones aside from gopls ?

**Comment by teto** [comment]
That spec is weird imo https://microsoft.github.io/language-server-protocol/specification#textEditArray, there should be no mutiple edits with the same position, like the server could just send one edit request. 

> I am unsure if this is a functionality that is desired. I interpret the spec to actually make remove followed by insert an invalid operation, but it seems like a reasonable thing to do

I would rather stick to the spec. Can't this be acheived with a replace edit instead ? Do you know some doc that justify gopls behavior ?

File: test/functional/plugin/lsp_spec.lua (hunk 1 above)
Comment (ndreas) [reply test/functional/plugin/lsp_spec.lua:880]: None that I'm aware of, but I suspect there might be more than gopls. Should I change this comment? I don't want it to become stale if more are discovered or gopls changes.

**Comment by ndreas** [comment]
> That spec is weird imo https://microsoft.github.io/language-server-protocol/specification#textEditArray, there should be no mutiple edits with the same position, like the server could just send one edit request.
> 
> > I am unsure if this is a functionality that is desired. I interpret the spec to actually make remove followed by insert an invalid operation, but it seems like a reasonable thing to do
> 
> I would rather stick to the spec. Can't this be acheived with a replace edit instead ? Do you know some doc that justify gopls behavior ?

I agree that the spec is weird, but I think it makes sense that the array represents an ordered list of changes such that change 1 is applied before change 2. The weirdness is that when you have the same insert, then that order actually changes.

I have not found any documentation, but I found that Microsoft's library for an LSP client for Node supports both cases:

Multiple inserts with the same position: https://github.com/microsoft/vscode-languageserver-node/blob/master/textDocument/src/test/edits.test.ts#L19

Replace followed by insert: https://github.com/microsoft/vscode-languageserver-node/blob/master/textDocument/src/test/edits.test.ts#L27

One thing I noticed in the test though was that the insert did not have the same start position as the replace, but rather the end position of the change. I think I might have mixed up the start and end position when I did #12434. I'm going to investigate this further and make sure that the spec is followed. I'll amend this PR with the changes and redo the title and description.

**Comment by ndreas** [comment]
This is so weird. I just did a revert on #12434, and everything just works. However, when I build neovim using a revision before #12434, it behaves weirdly.

Anyways, I'll close this, and open a new PR with the revert.

**Pull request closed without merging**